src/Chapter2.hs

```diff
@@ -727,7 +772,12 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x:xs) = concat $ repList (x:xs)
+    where
+      repList :: [Int] -> [[Int]]
+      repList [] = []
+      repList (y:ys) = map (replicate y) [y] ++ repList ys
```

Comment:
This is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `repList` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function, which you rightfully do!

For `repList` you can use standard function that does that – `replicate`.

Reply:
I cleaned up the clutter in the following code.

```haskell
smartReplicate :: [Int] -> [Int]
smartReplicate [] = []
smartReplicate (x:xs) = concat $ repList (x:xs)
    where
      repList (y:ys) = map (replicate y) [y] ++ repList ys
```

> For `repList` you can use standard function that does that – `replicate`.                  
I'm bit confused here - I've used `replicate` in `repList`. Can you please elaborate on this? 🙂

Reply:
Instead of helper function with a number of different actions, you could inline it into one, and you'll see that you don't need custom `repList`, but the standard `replicate` instead:

```haskell
smartReplicate xs = concat $ map (\x -> replicate x x) xs
```

And from here, you can use `concatMap` and eta-reduce as well 🙂
```haskell
smartReplicate = concatMap (\x -> replicate x x)
```
 Hope it makes more sense now 🙂 Sorry for the confusion!

Reply:
Oh Wow! I was mistaking the second argument of `replicate` for `[a]` rather than `a`. 😥 

That is a cool use of the lambda function. Taking notes ✍🏼. Thank You! 🙂